contracts/UpkeepRegistrationRequests.sol

```diff
@@ -134,23 +149,42 @@ contract UpkeepRegistrationRequests is Owned {
         uint32 gasLimit,
         address adminAddress,
         bytes calldata checkData,
-        uint96 amount,
         bytes32 hash
     )
       external
       onlyOwner()
     {
+        PendingRequest memory request = s_pendingRequests[hash];
+        require(request.admin != address(0), "request not found");
+        bytes32 expectedHash = keccak256(abi.encode(upkeepContract, gasLimit, adminAddress, checkData));
+        require(hash == expectedHash, "hash and payload do not match");
+        delete s_pendingRequests[hash];
         _approve(
             name,
             upkeepContract,
             gasLimit,
             adminAddress,
             checkData,
-            amount,
+            request.balance,
             hash
         );
     }
 
+    /**
+     * @notice cancel will remove a registration request and return the refunds to the msg.sender
+     * @param hash the request hash
+     */
+    function cancel(
+        bytes32 hash
+    )
+      external
+    {
+        PendingRequest memory request = s_pendingRequests[hash];
+        require(request.admin == msg.sender, "only admin can cancel");
```

Comment:
I think we should make this callable by the owner too, that way they can return funds instead of forcing the user to withdraw and implicitly rejecting by not accepting.

Reply:
@se3000 if we make this callable by the owner, then we'll want to add a `withdraw()` function too, correct? Otherwise this becomes push, instead of a pull.

Reply:
AFAIK the ["pull over push"](https://consensys.github.io/smart-contract-best-practices/recommendations/#favor-pull-over-push-for-external-calls) [pattern](https://fravoll.github.io/solidity-patterns/pull_over_push.html) focuses on sending Ether because that hands over execution control to the target contract and is less predictable about gas. I think this would be ok without withdrawal, but allowing the owner to call is not strictly necessary either.

Reply:
:+1: